dynamic_programming/catalan_numbers.nim

```diff
@@ -20,9 +21,8 @@ func catalanNumbersRecursive(index: Natural): Positive =
   ## efficient.
   if index < 2:
     return 1
-  for i in 0 ..< index:
-    result += catalanNumbersRecursive(i) *
-     catalanNumbersRecursive(index - i - 1)
+  return toSeq(0..<index).mapIt(catalanNumbersRecursive(it) *
+   catalanNumbersRecursive(index - it - 1)).foldl(a + b)
```

Comment:
What you've done here is instead of just solving the issue with an explicitly initialized variable, you've introduced a new dependency and use three new routines from it, two of which allocate a whole new sequence on each use, and all of it happens inside a recursive function.

```suggestion
  if index < 2:
    return 1
  var n: Natural = 0
  for i in 0 ..< index:
    n += catalanNumbersRecursive(i) * catalanNumbersRecursive(index - i - 1)
  n
```
(new variable `n` is necessary here as we can't init `result` to 0 as it's automatically typed as `Positive`)

IMO, users should be discouraged from using the "functional" templates from sequtils until they are fixed to compose properly. They are ok for most uses when not chained, but their behaviour is not exactly intuitive, especially if you forget they are in **seq**utils for a reason. They look innocent enough to be appealing but in this particular case (and many others) will lead to computational explosion.

In real-world practice users are better suited by [zero-functional](https://github.com/zero-functional/zero-functional) or [itermacros from beef331/slicerator](https://github.com/beef331/slicerator/blob/master/src/slicerator/itermacros.nim).

Reply:
@ZoomRmc thank you for this remark. I assumed that `seq` is a kind of _generator_ (i.e. it does not store the entire array).